include/material/norsand.tcc

```diff
@@ -0,0 +1,589 @@
+  double M_theta = Mtc_ - pow(Mtc_, 2) / (3. + Mtc_) * cos_lode_angle;
+  (*state_vars)["M_theta"] = M_theta;
+
+  return true;
+}
+
+//! Compute state parameters
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_state_variables(const Vector6d& stress,
+                                                 const Vector6d& dstrain,
+                                                 mpm::dense_map* state_vars) {
+
+  // Get state variables
+  const double mean_p = (*state_vars).at("p");
+  const double deviatoric_q = (*state_vars).at("q");
+  const double M_theta = (*state_vars).at("M_theta");
+  const double p_cohesion = (*state_vars).at("p_cohesion");
+
+  // Compute and update pressure image
+  double p_image;
+  if (bond_model_) {
+    p_image =
+        (mean_p + p_cohesion) *
+            pow(((1 - N_ / M_theta * deviatoric_q / (mean_p + p_cohesion)) /
+                 (1 - N_)),
+                ((N_ - 1) / N_)) -
+        p_cohesion;
+  } else {
+    p_image =
+        mean_p * pow(((1 - N_ / M_theta * deviatoric_q / mean_p) / (1 - N_)),
+                     ((N_ - 1) / N_));
+  }
+
+  (*state_vars).at("p_image") = p_image;
+
+  // Compute and update void ratio image
+  double e_image;
+  if (bond_model_) {
+    e_image = e_max_ - (e_max_ - e_min_) /
+                           log(crushing_pressure_ / (p_image + p_cohesion));
+  } else {
+    e_image = e_max_ - (e_max_ - e_min_) / log(crushing_pressure_ / p_image);
+  }
+  (*state_vars).at("e_image") = e_image;
+
+  // Update void ratio
+  // Note that dstrain is in tension positive - depsv = de / (1 + e)
+  double dvolumetric_strain = dstrain(0) + dstrain(1) + dstrain(2);
+  double void_ratio = (*state_vars).at("void_ratio") -
+                      (1 + (*state_vars).at("void_ratio")) * dvolumetric_strain;
+  (*state_vars).at("void_ratio") = void_ratio;
+
+  // Compute and update psi
+  double psi_image = void_ratio - e_image;
+  (*state_vars).at("psi_image") = psi_image;
+
+  return true;
+}
+
+//! Compute elastic tensor
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_p_cohesion(mpm::dense_map* state_vars) {
+
+  // Compute current zeta
+  double zeta = exp(-m_ * (*state_vars).at("epds"));
+
+  if (zeta > 1.)
+    (*state_vars).at("zeta") = 1.;
+  else if (zeta < 1.0E-15)
+    (*state_vars).at("zeta") = 0.;
+  else
+    (*state_vars).at("zeta") = zeta;
+
+  // Update p_cohesion
+  double p_cohesion = p_cohesion_initial_ * (*state_vars).at("zeta");
+  (*state_vars).at("p_cohesion") = p_cohesion;
+
+  return true;
+}
+
+//! Compute yield function and yield state
+template <unsigned Tdim>
+typename mpm::NorSand<Tdim>::FailureState
+    mpm::NorSand<Tdim>::compute_yield_state(double* yield_function,
+                                            const Vector6d& stress,
+                                            const mpm::dense_map* state_vars) {
+
+  // Get stress invariants
+  // Compute mean stress p
+  double mean_p = (stress(0) + stress(1) + stress(2)) / 3.;
```

Comment:
I saw you code duplicate the computation of p and q, I think it should be made to at least a function.

Reply:
Agree, but there are two parts here: (1) computing, (2) storing as state-variables. I want to use the `compute_stress_invariant` here but I don't want to update state-variables. 

Another way is to have a function to get p and q in which `compute_stress_invariant` and `compute_yield_state` call. But I am not sure what's the efficiency of doing it over a few lines of simple computations where we have to pass vectors and maps to do simple computation. @kks32, any comment?

I think if we want to make a function, we should make a state-variables update function too. It will make things cleaner.

Reply:
@ezrayst it won't increase in efficiency, just code maintenance. The more you have code duplication, the more your code prone to bugs. Anyway, we can refactor later as discussed.

Reply:
@ezrayst I agree with @bodhinandach have functions that compute `mean_p`, `j2`, `j3` etc so we don't have so many lines of code doing the same operations

Reply:
Will refactor just for `mean_p`, `j2`, `j3`, and `q` for now.

Reply:
The idea is to have less repetition